src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,599 @@
+        log.debug("key: {0}".format(key))
+        if metadata is None:
+            metadata = dict()
+        elif not isinstance(metadata, dict):
+            raise TypeError('*metadata*: expected dict or None, got %s' % type(metadata))
+
+        f = self._lookup_file(key, self.folder)
+        if f is not None:
+            log.debug("Object {0} already exist!".format(key))
+            self._delete_by_id(f['id'])
+        body = {
+            'name': key,
+            'mimeType': Backend.MIME_TYPE_BINARY,
+            'parents': [ self.folder['id'] ],
+            'properties': self._encode_metadata(metadata),
+        }
+        log.debug("metadata: {0}, body: {1}".format(metadata, body))
+        return ObjectW(self.service,self, key,body)
+
+    @prepend_ancestor_docstring
+    def clear(self):
+        log.debug("")
+        for f in self._list_files(self.folder, "id"):
+            self._delete_by_id(f['id'])            
+    '''
+    @copy_ancestor_docstring
+    def delete_multi(self, keys, force=False):
+        log.debug('started with %s', keys)
+
+        while len(keys) > 0:
+            tmp = keys[:MAX_KEYS]
+            self._delete_multi(tmp, force=force)
+            keys[:MAX_KEYS] = tmp
+
+    def _delete_file_async(request_id, response, exception):
+        if exception is not None:
+            log.debug('Object %s deleted successfully',request_id)
+            delete_map_state[request_id]="ok"
+        else:
+            log.warning('an error occur trying to delete %s',key)
+            delete_map_state[request_id]=exception
+        
+    def _is_batch_completed(self,dictionary):
+        ocurredException= None
+        for key, state in dictionary.items():
+            if state == "pending":
+                return False 
+            elif isinstance(state,Exception):                
+                ocurredException= state
+
+        #Only raise Exceptions when all request has been procesed
+        if ocurredException:
+            raise ocurredException
+
+    @retry
+    def _delete_multi(self, keys, force=False):
+        batch = self.service.new_batch_http_request(callback=_delete_file_async)
+        delete_map_state = dict()
+        for f in self._list_files(self.folder, "id,name"):
+            if name in keys:
+                delete_map_state[f[name]]="pending"
+                batch.add(request_id=name,request=self.service.files().delete(fileId=f["id"]))
+
+        batch.execute()
+        while self.is_batch_completed(deleteMapState)==False:
+            time.sleep(0.2)
+    '''
+    
+
+
+
+    @retry
+    @copy_ancestor_docstring
+    def delete(self, key, force=False, is_retry=False):
+        log.debug("key: {0}".format(key))
+        query = "name = '{0}'".format(self._escape_string(key))
+        found = False
+        for f in self._list_files(self.folder, "id", query):
+            found = True
+            self._delete_by_id(f['id'])
+        if not found:
+            if force or is_retry:
```

Comment:
Why ignore the error on retry?

Reply:
gdrive not always works "in sync" , some times you will get an error deleting a file but in the next retry request you find out the file doesn't exist anymore, it's weird, I add this condition to avoid s3ql crash when this kind of things happen.
When google scale up/down their infrastructure is common get error 500(it's in their documentation) I guess when this happens you could have an error 500 but the request is procesed correctly.

Reply:
Alright. Could you please put this into the code as a comment?